cis194/week1/daniel-deng/src/CIS194/Homework01/Exercise02.hs

```diff
@@ -1,9 +1,8 @@
 module CIS194.Homework01.Exercise02 (doubleEveryOther) where
 
-doubleEveryOtherFromLeft :: [Integer] -> [Integer]
-doubleEveryOtherFromLeft []                      = []
-doubleEveryOtherFromLeft [element]               = [element]
-doubleEveryOtherFromLeft (first : second : rest) = [first, second * 2] ++ doubleEveryOtherFromLeft rest
+import Data.Bool
 
 doubleEveryOther :: [Integer] -> [Integer]
-doubleEveryOther = reverse . doubleEveryOtherFromLeft . reverse
+doubleEveryOther = foldr f []
+                 where
+                   f x = flip (:) <*> bool x (x *2) . odd . length
```

Comment:
@shineli1984 which one do you reckon is better? Is that subjective or there's a rule as a guidance?

```suggestion
                   f x = (:) =<< bool x (x *2) . odd . length
```

Reply:
This actually turns the function quadratic because `length` has to go through the whole list.

So... I like the original version better

Reply:
> This actually turns the function quadratic because `length` has to go through the whole list.
> 
> So... I like the original version better

That depends on how Haskell runtime works on list length, worst case is O(n).
How about this anyway:

```haskell
doubleEveryOther :: [Integer] -> [Integer]
doubleEveryOther = snd . foldr f (0, [])
                 where
                   f x (c, xs) = (c + 1, newVal x c : xs)
                   newVal x    = bool x (x * 2) . odd
```